### DISCUSSION TITLE
Enum (and other?) custom sorting

### DISCUSSION THREAD

**Discussion by evantrimboli** [mikro-orm/mikro-orm#2178] [Ideas]
Using the following structure below, it's difficult to sort by priority since it will just sort on the natural string order. Ideally, I would be able to specify the order somehow, for example `defaultOrder`. It wouldn't need to be restricted to enums, it could obviously apply to string or other number types.

```ts
enum Priority {
  Low = 'low',
  Medium = 'medium',
  High = 'high'
}

@Entity()
class Task {
  @PrimaryKey()
  id!: number

  @Enum({ defaultOrder: [Priority.Low, Priority.Medium, Priority.High] })
  priority!: Priority
}
```

If I sorted by priority then, it could generate something like:

```sql
ORDER BY (CASE 
  WHEN "e0"."priority" = 'low' THEN 0
  WHEN "e0"."priority" = 'medium' THEN 1
  WHEN "e0"."priority" = 'high' THEN 2
  ELSE null
) ASC
```

Possible workarounds:
- Use a numeric enum -> Downsides are that the current schema generation doesn't limit put a constraint on the values, + the values aren't immediately obvious when viewing the data.
- Use "native" DB enums - These do implement a natural sort order, however they aren't cross platform and modifying them is a pain.
- Use query builder and write custom sorting - Try to avoid this where possible

Is this something you would be interested in supporting? If so I wouldn't mind having a stab at it if you could provide a bit of direction of:
- The API that suits
- At which point to implement it (I looked into the code and it seems it passes ordering through a few levels).

**Comment by B4nan** [comment]
I am not opposed to support this, we just need to think about the naming. Will get back to you when I have some time.

**Reply by evantrimboli** [reply]
Little bump on this, starting to get to a point where I need this functionality in my project.

Happy to look into an implementation once we settle on an API.

**Reply by B4nan** [reply]
I guess I am fine with what you proposed, we can iterate on specifics once you have something working, naming is not important at this stage either.

Maybe the `else null` should be rather `else 3` so when the value is not known, it will go always last?

I am not fully sure if it should go to a new decorator option, we already have `orderBy` there which is exactly that - the default ordering for given property (it is probably available only on collections now). Instead we should allow this feature in the existing `orderBy` option, as well as in the EntityManager and QueryBuilder APIs (it is all basically the same, handled by QB in SQL drivers, we should probably address this in mongo too, or at least throw when it gets detected). Maybe we could have a new operator for this:

```ts
@Enum({ orderBy: { $custom: [Priority.Low, Priority.Medium, Priority.High] } })
priority!: Priority
```

```ts
qb.select('*').orderBy({ $custom: [Priority.Low, Priority.Medium, Priority.High] })
```

```ts
em.find(Task, {}, { orderBy: { $custom: [Priority.Low, Priority.Medium, Priority.High] } })
```

**Reply by evantrimboli** [reply]
> Maybe the else null should be rather else 3 so when the value is not known, it will go always last?

Not sure about the `else null`, since you can choose whether you want to sort nulls first or not. Either way it's probably not ideal.

> I am not fully sure if it should go to a new decorator option

In my case I will never want the order to change, so if I say order by priority, I would want it to always order in that fashion without needing to specify it on the qb/find options. It somewhat defeats the purpose of the feature if I have to repeat it there.

For the qb/find options I think it needs to be extended a bit, so you could specify multiple fields + and the direction, something like:

```
{
  orderBy: {
    priority1: { 
      $custom: {
        values: [Priority.Low, Priority.Medium, Priority.High],
        direction: 'ASC'
      }
    },
    status: { 
      $custom: {
        values: [Status.Pending, Status.Progress, Status.Done],
        direction: 'DESC'
      }
    }
  }
}
```

In theory you could cater for the direction by virtue of setting the values, but if you're consuming the direction from elsewhere, it seems like it would be necessary to allow it to be appended.

**Reply by B4nan** [reply]
Oh I see, so you are suggesting a different approach, basically to replace `order by my_enum asc` with that order/case for given property - that way it makes sense and we will end up with simpler API. 

But then we should definitely find a different name than `defaultOrder`, to me it is really confusing given we have `orderBy` that works as the default order by for given property (for relations only and it automatically adds the order to the query when such relation is populated). Here you want to specify how ordering will work.

Maybe `customOrder` would fit better. 

>  I would want it to always order in that fashion without needing to specify it on the qb/find options.

I would very much like to not do that, not just because you specify how a column should handle ordering - that itself does not mean it should be always applied. We should have two options for this, your new one for specifying how to order, and the one I am mentioning that already works for collections, so simple `orderBy`. You should be able to use both at the same time to automatically apply it, but it feels like a wrong default to me, to merge those to a single option.

So for your use case where you want to also enable it by default, you would have something like:

```ts
@Enum({ customOrder: [Priority.Low, Priority.Medium, Priority.High], orderBy: QueryOrder.ASC })
priority!: Priority
```

I'd say we want values not defined in that array as always at the end, but we can iterate on that once we have some tests and we see how it behaves accross different drivers.

For EM/QB, things would work without any changes, no need for new operators, you'd just order by that column, specify direction as usual. 

```ts
qb.orderBy({ priority: QueryOrder.DESC })
em.find(Task, {}, { orderBy: { priority: QueryOrder.DESC } })
```

Btw this feature should work for numeric enums too, or better say for any string/number/bool column, not just enums.

**Reply by evantrimboli** [reply]
Yeah, I think we're on the same page, when I say "always order in that fashion", I mean whenever that property is used in the `orderBy`, it should generate the `case when` part for the query.
